src/variables.jl

```diff
@@ -22,10 +22,21 @@ struct ScalarVariable{S, T, U, V} <: AbstractVariable
     info::VariableInfo{S, T, U, V}
 end
 
-#############################################################################
-# VariableRef
-# Holds a reference to the model and the corresponding MOI.VariableIndex.
-struct VariableRef <: AbstractJuMPScalar
+"""
+    AbstractVariableRef
+
+Variable returned by [`addvariable`](@ref). Affine (resp. quadratic) operations with variables of type `V<:AbstractVariableRef` and coefficients of type `T` create a `GenericAffExpr{T,V}` (resp. `GenericQuadExpr{T,V}`).
+"""
+abstract type AbstractVariableRef <: AbstractJuMPScalar end
+
+variablereftype(v::AbstractVariableRef) = typeof(v)
```

Comment:
What's the point of this method?

Reply:
GenericAffExpr and GenericQuadExpr both work on arbitrary coefficient type and AbstractVariableRef type.
To write code that works for any `VariableRef`, it is often useful to get the variable ref type of an object that can be either an `AbstractVariableRef` or a `GenericAffExpr` or a `GenericQuadExpr`, e.g.
https://github.com/JuliaOpt/JuMP.jl/blob/151c75fba8e7a2f8b36e96aa29909441b06e9207/src/operators.jl#L73-L83